dashboard/lib/controllers/run.js

```diff
@@ -1,83 +1,98 @@
-        return runs.clear();
+        if (!run)
+            throw new Error('Unable to save run data. Incoherent schema.');
     },
 
     pause: async (id) => {
         if (!id) throw new TypeError('Invalid id');
-        const run = await api.findOne(id);
+        const run = await Run.findOne({ _id: id });
 
         if (!run) throw new Error('Run not found.');
-        run.setPaused();
+
+        run.status = 'paused';
+        await run.save();
     },
 
     resume: async (id) => {
         if (!id) throw new TypeError('Invalid id');
-        const run = await api.findOne(id);
+        const run = await Run.findOne({ _id: id });
 
         if (!run) throw new Error('Run not found.');
-        run.setActive();
+
+        run.status = 'active';
+        await run.save();
     },
 
     abort: async (id) => {
         if (!id) throw new TypeError('Invalid id');
-        const run = await api.findOne(id);
+        const run = await Run.findOne({ _id: id });
 
         if (!run) throw new Error('Run not found.');
-        run.setAborted();
+
+        run.status = 'aborted';
+        run.endTime = Date.now();
+        await run.save();
     },
 
     done: async (id) => {
         if (!id) throw new TypeError('Invalid id');
-        const run = await api.findOne(id);
+        const run = await Run.findOne({ _id: id });
 
         if (!run) throw new Error('Run not found.');
-        run.setFinished();
+
+        if (run.status === 'aborted') return;
+
+        run.status = 'finished';
+        run.endTime = Date.now();
+        await run.save();
     },
 
     interrupt: async (id) => {
         if (!id) throw new TypeError('Invalid id');
-        const run = await api.findOne(id);
+        const run = await Run.findOne({ _id: id });
 
         if (!run) throw new Error('Run not found.');
-        run.setInterrupted();
+
+        run.status = 'interrupted';
+        await run.save();
     },
 
-    addEvent: async (data) => {
-        if (!data.id) throw new TypeError('Invalid id');
-        const run = await api.findOne(data.id);
+    saveEvent: async (data) => {
+        if (!data.runId) throw new TypeError('Invalid parent id');
 
-        if (!run) throw new Error('Run not found.');
-        run.addEvent(data);
+        const event = Event.create(data);
+        if (!event) throw new Error('Invalid event type');
+        await event.save();
     },
 
-    addStats: async (data) => {
-        if (!data.id) throw new TypeError('Invalid id');
-        const run = await api.findOne(data.id);
+    saveStats: async (data) => {
+        if (!data.runId) throw new TypeError('Invalid parent id');
 
-        if (!run) throw new Error('Run not found.');
-        run.addRunStats(data);
+        const stat = RunStat.create(data);
+        if (!stat) throw new Error('Invalid stat type');
+
+        await stat.save();
+    },
+
+    addEvent: async (data) => {
+        const batch = new BatchQueue(api.saveEvent);
```

Comment:
We are creating a new queue every time this function is called. Is this expected? 🤔

Reply:
Yes. The queue should be initialised once. Either when the file is imported or when the first call is made.

Also, do note that:
- If multiple actions depend on each other, they should use a single queue.
- If actions are independent but depend on multiple calls of itself, it should use a single queue.
- If actions are not dependent on multiple calls of itself either, it doesn't need a queue.